README.md
# README

本项目是专为清华大学本科生工作助理（本科生辅导员）设计的 GPA 计算与排序工具。

## 使用须知

### 使用方法

#### 方案1：下载发行版直接运行（推荐）

1. 在 `Releases` 界面下载最新发行版  `GPAsorter-THU.exe`
2. 仔细阅读输入输出，按要求导出成绩表格原始文件
3. 打开命令行或双击左键直接运行已编译好的可执行程序

#### 方案2：个性化定制后编译运行

* 开发环境要求
  * python 3.7+
  * pyinstaller
* 脚本编译运行：下载源码至 `your_path` 文件夹，在 `your_path/GPAsorter-THU` 中打开命令行运行 `Shell` 脚本 `run.sh` 进行编译，编译好的可执行程序将被置于 `dist` 文件夹内并自动直接运行。如在 git bash 界面下，可运行如下命令
```
PS your_path/GPAsorter-THU
$ sh run.sh
```
其中，`your_path` 是指运行 `git clone` 下载此项目时所在的文件路径。

### 输入输出

1. 输入文件要求为 `.txt` 文本文件，需要将从信息门户“管理”界面导出的成绩表格中的必要数据粘贴到该文本文件，至少包括学号、姓名、教学班级、课程号、课程名、成绩、绩点成绩、考试时间、学分、课程属性、重修补考标志、特殊课程标记等信息。
  * 注：下载时会出现两项名称为“特殊课程标记”，请勾选值为“第一学位课程”的那一项（一般该项位置靠后）。
2. 运行程序时需要用户在命令行键入输入/输出文件的名称（不带后缀名，均默认为 `.txt` 文本文件）；输入/输出文件默认位于可执行程序所在文件夹，否则需要另行前缀相对路径（以**可执行程序的所在路径**为基准）。

## 功能特性

* 适合清华成绩体系，自动识别各种课程类型
  * 经过多次手动验证，所得数据结果有保证，示例输入/输出文件见 `test` 文件夹
  * 加入数据预验证环节，帮助用户快速锁定问题，并提示输入数据必需类型
* 设置弹性选项，适应不同标准下的成绩计算需求
  * 允许选择文本文件采用的字符编码方案（默认选择 UTF-8 方案）
  * 允许选择是否在任选课中剔除双学位和辅修课程（默认选择剔除）
  * 允许选择是否在已重修通过科目中剔除挂科记录（默认选择剔除）
  * 允许选择在计算 GPA 时是否剔除已通过科目中的 P/F 课程（默认选择剔除）

## 致谢

* 感谢生命科学系李天奇学长的前期开发工作！


CHANGELOG.md
# CHANGELOG

## Before change

* Added － 这里记录新增加了哪些功能／接口
* Changed － 功能／接口变更
* Deprecated － 不建议使用的功能／接口，将来会删掉
* Removed － 之前不建议使用的功能／接口，这次真的删掉了
* Fixed － 这里记录解决了哪些问题
* Others － 这里记录性能优化和安全性增强等改进。


## v1.0 (2018/02/16)

* Init：适合清华成绩体系，自动识别各种课程类型
* Init：经过多次手动验证，所得数据结果有保证

## v2.0 (2020/01/30)

* Added：允许选择是否在任选课中剔除双学位和辅修课程
* Added：允许选择是否在已重修通过科目中剔除挂科记录
* Added：加入数据预验证环节，帮助用户快速锁定问题

## v2.1 (2020/05/13)

* Fixed：修正了必限学分为 0 时会报错的 bug

## v2.2 (2020/07/08)

* Added：允许选择文本文件采用的字符编码方案（依照 Windows 记事本对编码模式的命名方式）
* Added：允许选择在计算 GPA 时是否剔除已通过科目中的 P/F 课程
* Changed：将输入/输出文件的类型限制为 `.txt` 文本文件

## v2.3 (2020/09/30)

* Changed: 为弹性选项设置默认值（通过直接回车确认），增加使用便捷性
* Fixed：修正了体疗课程计算 GPA 时的有效绩点值（1.3 * 0.8 = 1.04）


GPAcalculator.py
#! /usr/bin/env python3
# -*- coding: UTF-8 -*-

import os

## Function: Calculate total credit
def cal_total_credit(course_credit,grade_point,total_credit,grade,flag_exclude_pass):
  if grade in ['EX','W','I','G']:
    return total_credit
  elif grade == 'P':
    if flag_exclude_pass == 'T':
      return total_credit
    else:
      total_credit_new = total_credit + int(course_credit)
      return total_credit_new
  elif grade == 'F':
    if grade_point == 'N/A':
      return total_credit
    else:
      total_credit_new = total_credit + int(course_credit)
      return total_credit_new
  else:
    total_credit_new = total_credit + int(course_credit)
    return total_credit_new


## Function: Calculate GPA (Grade Point Average)
def cal_GPA(course_credit,grade_point,grade_points,grade):
  if grade in ['EX','P','W','I','G']:
    return grade_points
  elif grade == 'F':
    if grade_point == 'N/A':
      return grade_points
    else:
      grade_points_new = grade_points + int(course_credit)*float(grade_point)
      return grade_points_new
  else:
    grade_points_new = grade_points + int(course_credit)*float(grade_point)
    return grade_points_new


## Function: Detect the input string variables
def input_var(leading_words, choiceA, choiceA_words, choiceB, choiceB_words):
  var_value = input(leading_words)
  while var_value != choiceA and var_value != choiceB:
    var_value = input('输入有误。'+leading_words)
  if var_value == choiceA:
    print(choiceA_words)
  elif var_value == choiceB:
    print(choiceB_words)
  return var_value


## Function: Data preprocessing
def preproc(input_file_name,char_encoding):
  ## Input Flags
  flag_exclude_double_degree = input_var('是否在任选课中剔除双学位和辅修课程：\
                          \n\t剔除请输入T，保留请输入F，以回车结尾\n', \
                          \
                          'T','已选择剔除双学位和辅修课程\n', \
                          \
                          'F','已选择保留双学位和辅修课程\n')
  
  flag_exclude_fail_revamp = input_var('是否剔除已重修通过科目中的挂科记录：\
                          \n\t剔除请输入T，保留请输入F，以回车结尾\n', \
                          \
                          'T','已选择剔除已重修通过科目中的挂科记录\n', \
                          \
                          'F','已选择保留已重修通过科目的挂科记录\n')
  
  ## Write to "__temp.txt"
  with open(input_file_name,'r', encoding=char_encoding) as inp, \
    open('__temp.txt','a', encoding=char_encoding) as oup:
    title = inp.readline()
    title_col = title.rstrip().split('\t')

    # Detect the data completeness
    try:
      student_id    = title_col.index('学号')
      student_name  = title_col.index('姓名')
      teach_class   = title_col.index('教学班级')
      course_name   = title_col.index('课程名')
      course_id     = title_col.index('课程号')
      grade         = title_col.index('成绩')
      grade_point   = title_col.index('绩点成绩')
      exam_time     = title_col.index('考试时间')
      course_credit = title_col.index('学分')
      course_type   = title_col.index('课程属性')
      revamp_mark   = title_col.index('重修补考标志')
      degree_type   = title_col.index('特殊课程标记')
    except:
      print('数据不完整。请将必要数据粘贴到txt文件中使用，并至少包含如下信息：\
         \n\t学号，姓名，教学班级，课程号，课程名，成绩，绩点成绩，考试时间，\
         \n\t学分，课程属性，重修补考标志，特殊课程标记等\n')
      os.system('pause')
      exit()
    oup.write(title)
    lines = inp.readlines()
    total_line = []

    # Establish the dictionary: Point "student_id+course_id" to the last exam_time
    revamp_dict = {}
    for j in lines:
      i = j.rstrip().split('\t')
      total_line.append(j)
      if i[revamp_mark] =='重修':
        if i[student_id]+i[course_id] in revamp_dict.keys():
          if int(i[exam_time]) > int(revamp_dict[i[student_id]+i[course_id]]):
            revamp_dict[i[student_id]+i[course_id]]= i[exam_time]
        else:
          revamp_dict[i[student_id]+i[course_id]] = i[exam_time]
    
    # Delete Dual Degree Courses according to "student_id+course_id"
    final_line = []
    if flag_exclude_double_degree == 'F':
      final_line = total_line
    else:
      for i in total_line:
        j = i.rstrip().split('\t')
        if j[degree_type] == '第一学位课程':
          final_line.append(i)
    
    # Save all the passing courses and the last one in revamped courses
    final_line2 = []
    if flag_exclude_fail_revamp =='F':
      final_line2 = final_line
    else:
      for i in final_line:
        j = i.rstrip().split()
        if j[student_id]+j[course_id] not in revamp_dict.keys():
          final_line2.append(i)
        elif j[exam_time] == revamp_dict[j[student_id]+j[course_id]]:
          final_line2.append(i)
    for j in final_line2:    
      oup.write(j)


## Function: Process and output the GPA
def mainproc(output_file_name,char_encoding):
  # Input Flags
  flag_exclude_pass = input_var('是否在计算GPA时剔除已通过科目中的 P/F 课程：\
                        \n\t剔除请输入T（一般选择剔除），保留请输入F，以回车结尾\n', \
                        \
                        'T','已选择剔除已通过科目中的 P/F 课程\n', \
                        \
                        'F','已选择保留已通过科目中的 P/F 课程\n')
  
  # Read cleaned "__temp.txt" and calculate GPA
  with open('__temp.txt', 'r', encoding=char_encoding) as inp, \
    open(output_file_name, 'w', encoding=char_encoding) as oup:
    title = inp.readline()
    title_col = title.rstrip().split()

    student_id    = title_col.index('学号')
    student_name  = title_col.index('姓名')
    teach_class   = title_col.index('教学班级')
    course_name   = title_col.index('课程名')
    grade         = title_col.index('成绩')
    grade_point   = title_col.index('绩点成绩')
    course_credit = title_col.index('学分')
    course_type   = title_col.index('课程属性')
    revamp_mark   = title_col.index('重修补考标志')
    degree_type   = title_col.index('特殊课程标记')

    stu_course_list = inp.readlines()

    student_id_n   = []
    student_name_n = []
    teach_class_n  = []

    bixian_GPA             = []
    bixian_credits         = []
    bixian_grade_points    = []
    bixianren_GPA          = []
    bixianren_credits      = []
    bixianren_grade_points = []
    if flag_exclude_pass == 'T':
      bixianren_credits_incP = []
      string_exclude_pass = '（不计入P课程）'
    else:
      string_exclude_pass = '（计入P课程）'

    # Calculate credits and grade points
    j = 0
    for i in stu_course_list:
      i = i.rstrip().split('\t')
      if i[student_id] not in student_id_n:
        student_id_n.append(i[student_id])
        student_name_n.append(i[student_name])
        teach_class_n.append(i[teach_class])
        bixian_credits.append(0)
        bixian_grade_points.append(0)
        bixianren_credits.append(0)
        bixianren_grade_points.append(0)
        bixianren_credits_incP.append(0)
        j = j + 1
        print(str(j)+' '+student_name_n[j-1])
      poc = student_id_n.index(i[student_id])
      if '体疗' in i[course_name]:
        bixianren_credits[poc] = bixianren_credits[poc] + 1
        if flag_exclude_pass == 'T':
          bixianren_credits_incP[poc] = bixianren_credits_incP[poc] + 1
        bixianren_grade_points[poc] = bixianren_grade_points[poc] + 0.8
        print('含有体疗课程')
        continue        
      if i[course_type] in ['必修','限选']:
        bixian_credits[poc] = cal_total_credit(i[course_credit],i[grade_point],bixian_credits[poc],i[grade],flag_exclude_pass)
        bixian_grade_points[poc] = cal_GPA(i[course_credit],i[grade_point],bixian_grade_points[poc],i[grade])
      if flag_exclude_pass == 'T':
        bixianren_credits_incP[poc] = cal_total_credit(i[course_credit],i[grade_point],bixianren_credits_incP[poc],i[grade],'F')
      bixianren_credits[poc] = cal_total_credit(i[course_credit],i[grade_point],bixianren_credits[poc],i[grade],flag_exclude_pass)
      bixianren_grade_points[poc] = cal_GPA(i[course_credit],i[grade_point],bixianren_grade_points[poc],i[grade])
    print("总统计人数为："+str(j)+"\n")

    # Calculate GPA
    for g in range(0,j):
      if bixian_credits[g]==0:
        bixian_GPA.append(0)
      else:
        bixian_GPA.append(bixian_grade_points[g]/bixian_credits[g])
      if bixianren_credits[g]==0:
        bixianren_GPA.append(0)
      else:
        bixianren_GPA.append(bixianren_grade_points[g]/bixianren_credits[g])

    # Write to the output file
    if flag_exclude_pass == 'T':
      oup.write('学号\t姓名\t教学班级\t必限总学分'+string_exclude_pass+'\t必限总绩\t必限GPA\tRank-必限GPA'+\
                    '\t必限任总学分'+string_exclude_pass+'\t必限任总绩\t必限任GPA\tRank-必限任GPA'+\
                    '\t必限任总学分（计入P课程）\n')
      for g in range(0,j):
        oup.write(str(student_id_n[g])+'\t'+str(student_name_n[g])+'\t'+str(teach_class_n[g])+ \
            '\t'+str(bixian_credits[g])+'\t'+str(round(bixian_grade_points[g],3))+ \
            '\t'+str(round(bixian_GPA[g],3))+'\t'+str(sorted(bixian_GPA,reverse = True).index(bixian_GPA[g])+1)+ \
            '\t'+str(bixianren_credits[g])+'\t'+str(round(bixianren_grade_points[g],3))+ \
            '\t'+str(round(bixianren_GPA[g],3))+'\t'+str(sorted(bixianren_GPA,reverse = True).index(bixianren_GPA[g])+1)+ \
            '\t'+str(bixianren_credits_incP[g])+'\n')
    else:
      oup.write('学号\t姓名\t教学班级\t必限总学分'+string_exclude_pass+'\t必限总绩\t必限GPA'+string_exclude_pass+'\tRank-必限GPA'+\
                    '\t必限任总学分'+string_exclude_pass+'\t必限任总绩\t必限任GPA'+string_exclude_pass+'\tRank-必限任GPA\n')
      for g in range(0,j):
        oup.write(str(student_id_n[g])+'\t'+str(student_name_n[g])+'\t'+str(teach_class_n[g])+ \
            '\t'+str(bixian_credits[g])+'\t'+str(round(bixian_grade_points[g],3))+ \
            '\t'+str(round(bixian_GPA[g],3))+'\t'+str(sorted(bixian_GPA,reverse = True).index(bixian_GPA[g])+1)+ \
            '\t'+str(bixianren_credits[g])+'\t'+str(round(bixianren_grade_points[g],3))+ \
            '\t'+str(round(bixianren_GPA[g],3))+'\t'+str(sorted(bixianren_GPA,reverse = True).index(bixianren_GPA[g])+1)+ \
            '\n')
  
  os.remove('__temp.txt') 


## Main process
if __name__ == '__main__':
  print("亲爱的用户，欢迎您试用“李导帮你算成绩”程序2.2版本 \
      \n感谢您对本程序的支持 \
      \n如有程序运行错误，欢迎微信联系作者李天奇：litq_94 \
      \n一键生成成绩排名-尽享假期生活\n\n")
  input_file_name = input('\n请输入你要进行成绩排序的 txt 输入文件名称，回车键结束 \
              \n\t（不能带后缀名；输入文件默认位于可执行程序所在文件夹，也可选择另行前缀相对路径）\n')
  output_file_name = input('\n请输入你想要写入排序结果的 txt 输出文件名称，回车键结束 \
              \n\t（不能带后缀名；输出文件默认位于可执行程序所在文件夹，也可选择另行前缀相对路径）\n')
  char_encoding = input_var('\n请选择文本文件采用的字符编码方案（依照 Windows 记事本对编码模式的命名方式）：\
              \n\tUTF-8编码请输入utf-8，ANSI编码请输入ansi（为记事本默认存储编码），以回车结尾\n', \
              \
              'utf-8','已选择UTF-8编码方案\n', \
              \
              'ansi','已选择ANSI编码方案\n')
  preproc(input_file_name+'.txt',char_encoding)
  mainproc(output_file_name+'.txt',char_encoding)
  print("排名计算已结束，可以在 txt 输出文件中进行查看。\n 欢迎您下次使用。如果进一步需求，欢迎微信联系原作者李天奇：litq_94")
  os.system('pause')


GPAsorter-THU.py
#! /usr/bin/env python3
# -*- coding: UTF-8 -*-
"""GPA-sorter-THU
* course-related variables
  * course_credit：课程学分(Course Credit)
  * course_name：课程名(Course Name)
  * course_type：课程属性(Course Type)
  * course_id：课程号(Course ID)
  * course_order：课序号(Course ID Order)
  * bixian_*：必修+限选
  * bixianren_*：必修+限选+任选
* student-related variables
  * student_id：学号(Student ID)
  * student_name：学生姓名(Student Name)
  * teach_class：教学班级(Teaching Class)
  * exam_time：考试时间(Exam Time)
  * revamp_mark：重修补考标志(Revamp Mark)
  * degree_type：特殊课程标记(Degree Type)
  * grade：单一课程等级成绩(Grade)
  * grade_point: 单一课程绩点成绩(Grade Point)
  * *_credits：某一类型课程总学分(Total Credit)
  * *_grade_points：某一类型课程总学分绩(Total Grade Point)
  * *_GPA：某一类型课程平均学分绩(Grade Point Average)
"""

import os

def main():
  """main procedure
  """
  # Print leading words
  print("\n亲爱的用户，欢迎您试用“李导帮你算成绩”程序，感谢您对本程序的支持！ \
      \n\n如有程序运行错误，欢迎微信联系程序原作者或在 GitHub 项目提交新的 issue\
      \n- 程序原作者：李天奇（微信号 litq_94）\
      \n- 项目 GitHub 网址：https://github.com/huangyf15/GPAsorter-THU \
      \n\n--- 一键生成成绩排名-尽享愉快假期生活 --- \n")
  print("--------------------- Input information ---------------------")
  # Input necessary info
  input_file_name = input('\n- 请输入你要进行成绩排序的 txt 输入文件名称，回车键结束 \
              \n\t（不能带后缀名；输入文件默认位于可执行程序所在文件夹，也可选择另行前缀相对路径）\n')
  output_file_name = input('\n- 请输入你想要写入排序结果的 txt 输出文件名称，回车键结束 \
              \n\t（不能带后缀名；输出文件默认位于可执行程序所在文件夹，也可选择另行前缀相对路径）\n')
  char_encoding = input_var('\n- 请选择文本文件采用的字符编码方案（依照 Windows 记事本对编码模式的命名方式）：\
              \n\tUTF-8 编码请输入 utf-8，ANSI 编码请输入 ansi（为记事本默认存储编码），以回车结尾 \
              \n\t默认：UTF-8 编码，直接回车即可\n', \
              \
              'utf-8','已选择 UTF-8 编码方案\n', \
              \
              'ansi','已选择 ANSI 编码方案\n')
  # Calculate and sort
  preproc(input_file_name+'.txt',char_encoding)
  cal_and_sort(output_file_name+'.txt',char_encoding)
  print("排名计算已结束，可以在 txt 输出文件中进行查看。欢迎您下次使用！\
        \n\n如有进一步需求，欢迎微信联系原作者或在 GitHub 项目提交新的 issue：\
        \n- 程序原作者：李天奇（微信号 litq_94）\
        \n- 项目 GitHub 网址：https://github.com/huangyf15/GPAsorter-THU\n")


def cal_total_credit(course_credit,grade_point,total_credit,grade,flag_exclude_pass):
  """calculate total credit
  
  Returns: 
    total credit
  """
  if grade in ['EX','W','I','G']:
    return total_credit
  elif grade == 'P':
    if flag_exclude_pass == 'T':
      return total_credit
    else:
      total_credit_new = total_credit + int(course_credit)
      return total_credit_new
  elif grade == 'F':
    if grade_point == 'N/A':
      return total_credit
    else:
      total_credit_new = total_credit + int(course_credit)
      return total_credit_new
  else:
    total_credit_new = total_credit + int(course_credit)
    return total_credit_new


def cal_GPA(course_credit,grade_point,grade_points,grade):
  """calculate GPA (Grade Point Average)
  
  Returns: 
    total grade point, i.e. GPA
  """
  if grade in ['EX','P','W','I','G']:
    return grade_points
  elif grade == 'F':
    if grade_point == 'N/A':
      return grade_points
    else:
      grade_points_new = grade_points + int(course_credit)*float(grade_point)
      return grade_points_new
  else:
    grade_points_new = grade_points + int(course_credit)*float(grade_point)
    return grade_points_new


def input_var(leading_words, choiceA, choiceA_words, choiceB, choiceB_words):
  """detect the input string variables

  Returns:
    legal string value (choiceA or choiceB)
  """
  var_value = input(leading_words)
  while var_value != "" and var_value != choiceA and var_value != choiceB:
    var_value = input('输入有误。'+leading_words)
  if var_value == "":
    var_value = choiceA
  if var_value == choiceA:
    print(choiceA_words)
  elif var_value == choiceB:
    print(choiceB_words)
  return var_value


def preproc(input_file_name,char_encoding):
  """preprocess the data

  Write the preprocessed data to a temporary file.
  """
  # Print leading words
  print("----------------------- Preprocessing -----------------------")
  # Input Flags
  flag_exclude_double_degree = input_var('\n- 是否在任选课中剔除双学位和辅修课程：\
                          \n\t剔除请输入 T，保留请输入 F，以回车结尾 \
                          \n\t默认：T（剔除双学位和辅修课程），直接回车即可\n', \
                          \
                          'T','已选择剔除双学位和辅修课程\n', \
                          \
                          'F','已选择保留双学位和辅修课程\n')
  
  flag_exclude_fail_revamp = input_var('\n- 是否剔除已重修通过科目中的挂科记录：\
                          \n\t剔除请输入 T，保留请输入 F，以回车结尾 \
                          \n\t默认：T（剔除已重修通过科目中的挂科记录），直接回车即可\n', \
                          \
                          'T','已选择剔除已重修通过科目中的挂科记录\n', \
                          \
                          'F','已选择保留已重修通过科目的挂科记录\n')
  
  # Write to "__temp.txt"
  with open(input_file_name,'r', encoding=char_encoding) as inp, \
    open('__temp.txt','a', encoding=char_encoding) as oup:
    title = inp.readline()
    title_col = title.rstrip().split('\t')

    # Detect the data completeness
    try:
      student_id    = title_col.index('学号')
      student_name  = title_col.index('姓名')
      teach_class   = title_col.index('教学班级')
      course_name   = title_col.index('课程名')
      course_id     = title_col.index('课程号')
      grade         = title_col.index('成绩')
      grade_point   = title_col.index('绩点成绩')
      exam_time     = title_col.index('考试时间')
      course_credit = title_col.index('学分')
      course_type   = title_col.index('课程属性')
      revamp_mark   = title_col.index('重修补考标志')
      degree_type   = title_col.index('特殊课程标记')
    except:
      print('数据不完整。请将必要数据粘贴到txt文件中使用，并至少包含如下信息：\
         \n\t学号，姓名，教学班级，课程号，课程名，成绩，绩点成绩，考试时间，\
         \n\t学分，课程属性，重修补考标志，特殊课程标记等\n')
      os.system('pause')
      exit()
    oup.write(title)
    lines = inp.readlines()
    total_line = []

    # Establish the dictionary: Point "student_id+course_id" to the last exam_time
    revamp_dict = {}
    for j in lines:
      i = j.rstrip().split('\t')
      total_line.append(j)
      if i[revamp_mark] =='重修':
        if i[student_id]+i[course_id] in revamp_dict.keys():
          if int(i[exam_time]) > int(revamp_dict[i[student_id]+i[course_id]]):
            revamp_dict[i[student_id]+i[course_id]]= i[exam_time]
        else:
          revamp_dict[i[student_id]+i[course_id]] = i[exam_time]
    
    # Delete Dual Degree Courses according to "student_id+course_id"
    final_line = []
    if flag_exclude_double_degree == 'F':
      final_line = total_line
    else:
      for i in total_line:
        j = i.rstrip().split('\t')
        if j[degree_type] == '第一学位课程':
          final_line.append(i)
    
    # Save all the passing courses and the last one in revamped courses
    final_line2 = []
    if flag_exclude_fail_revamp =='F':
      final_line2 = final_line
    else:
      for i in final_line:
        j = i.rstrip().split()
        if j[student_id]+j[course_id] not in revamp_dict.keys():
          final_line2.append(i)
        elif j[exam_time] == revamp_dict[j[student_id]+j[course_id]]:
          final_line2.append(i)
    for j in final_line2:    
      oup.write(j)


def cal_and_sort(output_file_name,char_encoding):
  """calculate and output GPA

  Read data from the temporary file, calculate and sort students' GPA.
  """
  # Print leading words
  print("----------------------- Main process -----------------------")
  # Input Flags
  flag_exclude_pass = input_var('\n- 是否在计算 GPA 时剔除已通过科目中的 P/F 课程：\
                        \n\t剔除请输入 T（一般选择剔除），保留请输入 F，以回车结尾 \
                        \n\t默认：T（剔除已通过科目中的 P/F 课程），直接回车即可\n', \
                        \
                        'T','已选择剔除已通过科目中的 P/F 课程\n', \
                        \
                        'F','已选择保留已通过科目中的 P/F 课程\n')
  
  # Read cleaned "__temp.txt" and calculate GPA
  with open('__temp.txt', 'r', encoding=char_encoding) as inp, \
    open(output_file_name, 'w', encoding=char_encoding) as oup:
    title = inp.readline()
    title_col = title.rstrip().split()

    student_id    = title_col.index('学号')
    student_name  = title_col.index('姓名')
    teach_class   = title_col.index('教学班级')
    course_name   = title_col.index('课程名')
    grade         = title_col.index('成绩')
    grade_point   = title_col.index('绩点成绩')
    course_credit = title_col.index('学分')
    course_type   = title_col.index('课程属性')
    revamp_mark   = title_col.index('重修补考标志')
    degree_type   = title_col.index('特殊课程标记')

    stu_course_list = inp.readlines()

    student_id_n   = []
    student_name_n = []
    teach_class_n  = []

    bixian_GPA             = []
    bixian_credits         = []
    bixian_grade_points    = []
    bixianren_GPA          = []
    bixianren_credits      = []
    bixianren_grade_points = []
    if flag_exclude_pass == 'T':
      bixianren_credits_incP = []
      string_exclude_pass = '（不计入P课程）'
    else:
      string_exclude_pass = '（计入P课程）'

    # Calculate credits and grade points
    j = 0
    for i in stu_course_list:
      i = i.rstrip().split('\t')
      if i[student_id] not in student_id_n:
        student_id_n.append(i[student_id])
        student_name_n.append(i[student_name])
        teach_class_n.append(i[teach_class])
        bixian_credits.append(0)
        bixian_grade_points.append(0)
        bixianren_credits.append(0)
        bixianren_grade_points.append(0)
        bixianren_credits_incP.append(0)
        j = j + 1
        print(str(j)+' '+student_name_n[j-1])
      poc = student_id_n.index(i[student_id])
      if '体疗' in i[course_name]:
        bixianren_credits[poc] = bixianren_credits[poc] + 1
        if flag_exclude_pass == 'T':
          bixianren_credits_incP[poc] = bixianren_credits_incP[poc] + 1
        bixianren_grade_points[poc] = bixianren_grade_points[poc] + 0.8 * 1.3
        print('含有体疗课程')
        continue        
      if i[course_type] in ['必修','限选']:
        bixian_credits[poc] = cal_total_credit(i[course_credit],i[grade_point],bixian_credits[poc],i[grade],flag_exclude_pass)
        bixian_grade_points[poc] = cal_GPA(i[course_credit],i[grade_point],bixian_grade_points[poc],i[grade])
      if flag_exclude_pass == 'T':
        bixianren_credits_incP[poc] = cal_total_credit(i[course_credit],i[grade_point],bixianren_credits_incP[poc],i[grade],'F')
      bixianren_credits[poc] = cal_total_credit(i[course_credit],i[grade_point],bixianren_credits[poc],i[grade],flag_exclude_pass)
      bixianren_grade_points[poc] = cal_GPA(i[course_credit],i[grade_point],bixianren_grade_points[poc],i[grade])
    print("总统计人数为："+str(j)+"\n")

    # Calculate GPA
    for g in range(0,j):
      if bixian_credits[g]==0:
        bixian_GPA.append(0)
      else:
        bixian_GPA.append(bixian_grade_points[g]/bixian_credits[g])
      if bixianren_credits[g]==0:
        bixianren_GPA.append(0)
      else:
        bixianren_GPA.append(bixianren_grade_points[g]/bixianren_credits[g])

    # Write to the output file
    if flag_exclude_pass == 'T':
      oup.write('学号\t姓名\t教学班级\t必限总学分'+string_exclude_pass+'\t必限总绩\t必限GPA\tRank-必限GPA'+\
                    '\t必限任总学分'+string_exclude_pass+'\t必限任总绩\t必限任GPA\tRank-必限任GPA'+\
                    '\t必限任总学分（计入P课程）\n')
      for g in range(0,j):
        oup.write(str(student_id_n[g])+'\t'+str(student_name_n[g])+'\t'+str(teach_class_n[g])+ \
            '\t'+str(bixian_credits[g])+'\t'+str(round(bixian_grade_points[g],3))+ \
            '\t'+str(round(bixian_GPA[g],3))+'\t'+str(sorted(bixian_GPA,reverse = True).index(bixian_GPA[g])+1)+ \
            '\t'+str(bixianren_credits[g])+'\t'+str(round(bixianren_grade_points[g],3))+ \
            '\t'+str(round(bixianren_GPA[g],3))+'\t'+str(sorted(bixianren_GPA,reverse = True).index(bixianren_GPA[g])+1)+ \
            '\t'+str(bixianren_credits_incP[g])+'\n')
    else:
      oup.write('学号\t姓名\t教学班级\t必限总学分'+string_exclude_pass+'\t必限总绩\t必限GPA\tRank-必限GPA'+\
                    '\t必限任总学分'+string_exclude_pass+'\t必限任总绩\t必限任GPA\tRank-必限任GPA\n')
      for g in range(0,j):
        oup.write(str(student_id_n[g])+'\t'+str(student_name_n[g])+'\t'+str(teach_class_n[g])+ \
            '\t'+str(bixian_credits[g])+'\t'+str(round(bixian_grade_points[g],3))+ \
            '\t'+str(round(bixian_GPA[g],3))+'\t'+str(sorted(bixian_GPA,reverse = True).index(bixian_GPA[g])+1)+ \
            '\t'+str(bixianren_credits[g])+'\t'+str(round(bixianren_grade_points[g],3))+ \
            '\t'+str(round(bixianren_GPA[g],3))+'\t'+str(sorted(bixianren_GPA,reverse = True).index(bixianren_GPA[g])+1)+ \
            '\n')
  
  os.remove('__temp.txt') 


if __name__ == '__main__':
  main()
  os.system('pause')
